src/ast/ast.c

```diff
@@ -206,14 +171,331 @@ AST *AST_Build(cypher_parse_result_t *parse_result) {
+}
+
+// Map identifiers within an expression.
+static void _AST_MapExpression(AST *ast, const cypher_astnode_t *exp) {
+	cypher_astnode_type_t type = cypher_astnode_type(exp);
+
+	// In case of identifier.
+	if(type == CYPHER_AST_IDENTIFIER) {
+		const char *identifier_name = cypher_ast_identifier_get_name(exp);
+		_AST_UpdateRefMap(ast, identifier_name);
+	} else if(type == CYPHER_AST_PROPERTY_OPERATOR) {
+		// In case of property.
+		exp = cypher_ast_property_operator_get_expression(exp);
+		assert(cypher_astnode_type(exp) == CYPHER_AST_IDENTIFIER);
+		const char *identifier_name = cypher_ast_identifier_get_name(exp);
+		_AST_UpdateRefMap(ast, identifier_name);
+	} else {
+		// Recurse over children.
+		uint child_count = cypher_astnode_nchildren(exp);
+		for(uint i = 0; i < child_count; i++) {
+			const cypher_astnode_t *child = cypher_astnode_get_child(exp, i);
+			// Recursively continue mapping.
+			_AST_MapExpression(ast, child);
+		}
+	}
+}
+
+// Maps the RHS of "AS" projected entities.
+static inline void _AST_MapProjectionAlias(AST *ast, const cypher_astnode_t *projection) {
+	const cypher_astnode_t *ast_alias = cypher_ast_projection_get_alias(projection);
+	if(ast_alias == NULL) {
+		// The projection was not aliased, so the projection itself must be an identifier.
+		ast_alias = cypher_ast_projection_get_expression(projection);
+		assert(cypher_astnode_type(ast_alias) == CYPHER_AST_IDENTIFIER);
+	}
+	// WITH and RETURN projections are always either aliased or themselves identifiers.
+	const char *alias = cypher_ast_identifier_get_name(ast_alias);
+	_AST_UpdateRefMap(ast, alias);
+}
+
+// Adds referenced entities of ORDER BY clause.
+static void _AST_MapOrderByReferences(AST *ast, const cypher_astnode_t *order_by) {
+	uint count = cypher_ast_order_by_nitems(order_by);
+	// Go over each order by expression.
+	for(uint i = 0; i < count; i++) {
+		const cypher_astnode_t *item = cypher_ast_order_by_get_item(order_by, i);
+		const cypher_astnode_t *expression = cypher_ast_sort_item_get_expression(item);
+		_AST_MapExpression(ast, expression);
+	}
+}
+
+// Adds a node to the referenced entities rax, in case it has labels or properties (inline filter).
+static void _AST_MapReferencedNode(AST *ast, const cypher_astnode_t *node, bool include_all) {
+	if(include_all) { // WITH/RETURN * projection, include all user-defined aliases.
+		const cypher_astnode_t *identifier = cypher_ast_node_pattern_get_identifier(node);
+		const char *alias = cypher_ast_identifier_get_name(identifier);
+		// Include this alias if it is user-defined.
+		if(strncmp(alias, "anon_", 5)) _AST_UpdateRefMap(ast, alias);
+	}
+
+	const cypher_astnode_t *properties = cypher_ast_node_pattern_get_properties(node);
+	// A node with inlined filters is always referenced for the FilterTree.
+	// (In the case of a CREATE path, these are properties being set)
+	if(properties) {
+		const cypher_astnode_t *identifier = cypher_ast_node_pattern_get_identifier(node);
```

Comment:
At the AST level you're correct, but at the ExecutionPlan level `OpCreate` is considered to modify every entity it creates, even anonymous ones. (This logic can be changed later if we choose to not use the Record for storage of created entities.)

I don't really feel like it's worth making changes here, though, because this duality makes unnecessary AST references harmless, and we'd need to add additional interfaces to scan CREATE patterns and adequately capture references like `a` in:
`MATCH (a) CREATE (b {val: a.val} RETURN b)`

I'll make changes if you feel strongly, though!